enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java

```diff
@@ -270,4 +275,51 @@ public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleExcep
       throw new EnforcerRuleException("Failed to collect dependency " + ex.getMessage(), ex);
     }
   }
+
+  /**
+   * Returns {code Optional} describing the path from project root to a problematic artifact that
+   * caused {@link ArtifactTransferException}. An empty {@code Optional} if {@code exception} is not
+   * caused by {@link ArtifactTransferException}.
+   */
+  @VisibleForTesting
+  static Optional<String> formatDependencyPathInException(DependencyResolutionException exception) {
+    Throwable cause = exception.getCause();
+    while (cause != null) {
+      if (cause instanceof ArtifactTransferException) {
+        ArtifactTransferException artifactException = (ArtifactTransferException) cause;
+        return Optional.of(
+            findPaths(exception.getResult().getDependencyGraph(), artifactException.getArtifact()));
+      } else {
+        cause = cause.getCause();
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static String findPaths(DependencyNode root, Artifact artifact) {
+    ImmutableList.Builder<ImmutableList<DependencyNode>> result = ImmutableList.builder();
+
+    Deque<DependencyNode> stack = new ArrayDeque<>();
+    stack.addLast(root);
+    findArtifact(result, root, stack, artifact);
+
+    return result.build().stream()
```

Comment:
This feels excessively complex. I'm surprised you needed both Joiners and Collectors.joining. Possibly you do, but too much code is included in a single statement to make that apparent.